src/java/fr/paris/lutece/plugins/forms/web/admin/FormTransitionJspBean.java

```diff
@@ -565,15 +567,41 @@ private boolean validateTransition( Transition transition )
     private ReferenceList getTransitionTargetStepReferenceList( int nIdForm, int nIdStep )
     {
         ReferenceList listTransitionTargetSteps = new ReferenceList( );
-
+        List<Transition> transitions = TransitionHome.getTransitionsListFromForm( nIdForm );
         for ( ReferenceItem step : StepHome.getStepReferenceListByForm( nIdForm ) )
         {
-            if ( NumberUtils.toInt( step.getCode( ) ) != nIdStep )
+
+                if ( ( transitions.isEmpty( ) && NumberUtils.toInt( step.getCode( ) ) != nIdStep )
+                        || validateStep ( NumberUtils.toInt( step.getCode( ) ), nIdStep, transitions  ) )
             {
                 listTransitionTargetSteps.add( step );
             }
         }
         return listTransitionTargetSteps;
     }
 
+    /**
+     * @param nStepToValidate
+     *              the identifier of the step which will be add to the list of transitions steps
+     * @param nIdStep
+     *              the identifier of the current step
+     * @param transitions
+     *              the list of transitions
+     *              
+     * @return true if the step can be added to the list
+     */
+    private boolean validateStep( int nStepToValidate, int nIdStep, List<Transition> transitions )
+    {
+
+        return ( ( nStepToValidate != nIdStep
+                && transitions.stream( )
+                        .noneMatch( t -> nIdStep == t.getNextStep( ) && nStepToValidate == t.getFromStep( ) )
+                && transitions.stream( )
+                        .noneMatch( t -> nIdStep == t.getFromStep( ) && nStepToValidate == t.getNextStep( ) ) )
+                && ( ( transitions.stream( )
+                        .anyMatch( t -> t.getFromStep( ) != nIdStep && t.getNextStep( ) != nStepToValidate ) ) ||
+
+                        transitions.stream( ).noneMatch( t -> ( t.getFromStep( ) == nStepToValidate )
+                                && ( t.getNextStep( ) == nStepToValidate ) ) ) );
```

Comment:
The succession of lambda expressions makes the code difficult to read. It would have been better to do a single for loop on the list of transitions and check the different constraints on each one.